### ISSUE TITLE
cannot handly geometry type: GeometryCollection

### ISSUE THREAD

**Issue by gr2m** [mbloch/mapshaper#69]
First of, thanks for creating and sharing this amazing tool. I'm incredibly impressed! I reduced a 3.4 MB file down to a few kb, and one can barely tell the difference!

Unfortunately, reducing the file size also removed some of its shapes. My source file is this here:
https://gist.github.com/gr2m/8cee8d12f9b1afae4c4b

Mapshaper dosen't display some of the Somalia regions. If I try to upload one of the missing shapes directly, I get the  error

```
mapshaper.js:460 Uncaught Error: Layer is missing a valid geometry typemapshaper.js:460 errormapshaper.js:5720 MapShaper.getLayerBoundsmapshaper.js:5695 (anonymous function)mapshaper.js:5694 MapShaper.getDatasetBoundsmapshaper.js:13743 addDatamapshaper.js:11691 inputFileContentmapshaper.js:11644 readFile.reader.onloadFileReader (async)mapshaper.js:11649 readFilemapshaper.js:92 Utils.mapFiltermapshaper.js:855 Utils.forEachmapshaper.js:11632 readFilesmapshaper.js:11616 (anonymous function)mapshaper.js:2047 Handler.triggermapshaper.js:2100 EventDispatcher.dispatchEventmapshaper.js:5592 onchangemapshaper.js:2285 pageEvents.addEventListener.handler
```

Any chance I can fix / work around this?

**Comment by gr2m** [comment]
Here is a geojson file that causes the issue above:
https://gist.github.com/gr2m/f71ddb7f4d898a392975

**Comment by gr2m** [comment]
Just for reference, GeometryCollection is used because some regions have islands which are disconnected from the main regions. My workaround is to drop the islands and turn GeometryCollection into a Polygon, by picking the shape with the most coordinates. Here is the script I used:

```js
var json = require('./somalia.json');

json.features.forEach(function(feature/*, i*/) {
  if (feature.geometry.type === 'GeometryCollection') {
    console.log('- Converting GeometryCollection to Polygon for %s.', feature.properties.NAME_1);

    var numPolygonCoordinates = feature.geometry.geometries.map(function(g){
      return g.coordinates[0].length;
    });
    var max = Math.max.apply(Math, numPolygonCoordinates);
    var indexOfMaxValue = numPolygonCoordinates.indexOf(max);
    console.log('  found geometry with most coordinates (%d) at position (%d)', max, indexOfMaxValue);

    feature.geometry = feature.geometry.geometries[indexOfMaxValue];
  }
});

fs.writeFileSync('./somalia-polygons-only.json', JSON.stringify(json));
```

**Comment by mbloch** [comment]
Hi, I took a look at your GeoJSON files, and I see what's happening. 

Mapshaper can only handle GeoJSON that is organized in a simpler way than your source file. Most of the features in `somalia.geojson` are single-ring polygons; the GeoJSON types are nested like this: `Feature > Polygon`. Mapshaper does fine with these. There are also five multi-part features that are nested like this: `Feature > GeometryCollection > Polygon`. Mapshaper currently can't import geometry collections that are nested inside Feature objects. To preserve the islands, you can convert each GeometryCollection into a MultiPolygon object (see the GeoJSON spec if you're unsure about how to do this).

I'd like to improve Mapshaper's GeoJSON support so it can import files like yours... we can leave this issue open until that's implemented. (Also the program shouldn't throw an unhandled exception when this kind of thing occurs -- I'll fix that too).

Thanks :) mbloch

**Comment by mbloch** [comment]
Commit ac0c27bc84cdf71494182dc14 adds support for importing GeometryCollections that are nested inside of GeoJSON Features. After the update, you data file seems to load correctly.

Mapshaper is still unable to load GeoJSON files containing mixed point, line and polygon types, that's still on my 'TODO' list....

**Comment by mbloch** [comment]
Just published version 0.2.16, which includes the fix.
